text/0000-rewards.md

```diff
@@ -0,0 +1,250 @@
+
+We propose a simple approximate solution based upon computing medians across validators for used votes.
+
+0. In an epoch $e$, each validator $u$ counts of the number $\alpha_{u,v}$ of votes they *used* from each validator $v$, including themselves.  Any time a validator marks a candidate approved, they increment these counts appropriately. 
+
+1. After epoch $e$'s last block gets finalized, all validators of epoch $e$ submit an _approvals tally message_ `ApprovalsTallyMessage` that reveals their number $\alpha_{u,v}$ of useful approvals they saw from each validator $v$ on candidates that became available in epoch $n$.  We do not send $\alpha_{u,u}$ for tit-for-tat reasons discussed below, not for bias concerns.  We record these approvals tally messages on-chain.
+
+2. After some delay, we compute on-chain the median $\alpha_v := \textrm{median} \{ \alpha_{u,v} : u \}$ used approvals statements for each validator $v$. 
+
+As discussed in https://hackmd.io/@rgbPIkIdTwSICPuAq67Jbw/S1fHcvXSF we could compute these medians using the [on-line algorithm](https://www.quora.com/Is-there-an-online-algorithm-to-calculate-the-median-of-a-stream-of-numbers-if-stream-elements-can-be-added-or-removed-at-any-point?share=1) if substrate had a nice priority queue.
+
+We never achieve true consensus on approval checkers and their approval votes.  Yet, our approval assignment loop gives a rough concensus, under our Byzantine assumption and some synchrony assumption.  It then follows that miss-reporting by malicious validators should not appreciably alter the median $\alpha_v$ and hence rewards.  
+
+We never tally used approval assignments to candidate equivocations or other forks.  Any validator should always conclude whatever approval checks it begins, even on other forks, but we expect relay chain equivocations should be vanishingly rare, and sassafras should make forks uncommon.
+
+### Availability redistribution
+
+As approval checkers could easily perform useless checks, we shall reward availability providers for the availability chunks they provide that resulted in useful approval checks.  We enforce honesty using a tit-for-tat mechanism because chunk transfers are inherently subjective.
+
+An approval checker reconstructs the full parachain block by downloading distinct $f+1$ chunks from other validators, where at most $f$ validators are byzantine, out of the $n \ge 3 f + 1$ total validators.  In downloading chunks, validators prefer the $f+1$ systemic chunks over the non-systemic chunks, and prefer fetching from validators who already voted valid, like backing checkers.  It follows some validators should recieve credit for more than one chunk per candidate.  
```

Comment:
As above, fetching from backers is one of the two ways to fetch the systemic chunks.  We reward backers like the availability providing nodes in that case.

In principle if backer A provides like 330 systemic chunks, and then we need the other 4 or so chunks from the other backer and someone else, then backer A gets credited for 330, while the others get credited for what they provided.  Afaik it shouldn't matter if backer A gave us less than the full block, since they'd presumably provide it as chunks.

I suppose backers might not provide as chunks, but instead provide the parablock as a flat file, since chunks incur the overhead of the merkle proofs.  If this is better then that's fine, we can still pay the backer for f+1 chunks here.  I'm not sure this is better though.